lib/benchmark/benchmark.go

```diff
@@ -149,157 +157,133 @@ func (c *Config) Benchmark(ctx context.Context, tc *client.TeleportClient) (Resu
 	timeElapsed := false
 	start := time.Now()
-
 	for {
 		if c.MinimumWindow <= time.Since(start) {
 			timeElapsed = true
 		}
 		select {
-		case measure := <-responseC:
-			result.Histogram.RecordValue(int64(measure.End.Sub(measure.Start) / time.Millisecond))
-			results = append(results, measure)
-			if timeElapsed && len(results) >= c.MinimumMeasurements {
-				cancelWorkers()
+		case measure := <-resultC:
+			result.Histogram.RecordValue(int64(measure.End.Sub(measure.ResponseStart) / time.Millisecond))
+			result.RequestsOriginated++
+			if timeElapsed && result.RequestsOriginated >= c.MinimumMeasurements {
+				cancel()
 			}
 			if measure.Error != nil {
 				result.RequestsFailed++
 				result.LastError = measure.Error
 			}
-			result.RequestsOriginated++
 		case <-ctx.Done():
 			result.Duration = time.Since(start)
 			return result, nil
 		case <-statusTicker.C:
-			logrus.Infof("working... observations: %d", len(results))
+			logrus.Infof("working... current observation count: %d", result.RequestsOriginated)
 		}
+
 	}
 }
 
-func produceMeasures(ctx context.Context, rate int, c chan<- benchMeasure) {
-	interval := time.Duration(1 / float64(rate) * float64(time.Second))
-	ticker := time.NewTicker(interval)
-	defer ticker.Stop()
+type benchMeasure struct {
+	ResponseStart time.Time
+	End           time.Time
+	Error         error
+	ctx           context.Context
+	client        *client.TeleportClient
+	command       []string
+	interactive   bool
+}
+
+func run(ctx context.Context, request <-chan benchMeasure, done chan<- benchMeasure) {
 	for {
 		select {
-		case <-ticker.C:
-
-			measure := benchMeasure{
-				Start: time.Now(),
-			}
-			select {
-			case c <- measure:
-			case <-ctx.Done():
+		case m, ok := <-request:
+			if !ok {
 				return
 			}
+			go work(ctx, m, done)
 		case <-ctx.Done():
 			return
 		}
 	}
 }
 
-type benchMeasure struct {
-	Start           time.Time
-	End             time.Time
-	ThreadCompleted bool
-	ThreadID        int
-	Error           error
-}
-
-type benchmarkThread struct {
-	id          int
-	ctx         context.Context
-	client      *client.TeleportClient
-	command     []string
-	interactive bool
-	receiveC    chan benchMeasure
-	sendC       chan benchMeasure
+func work(ctx context.Context, m benchMeasure, send chan<- benchMeasure) {
+	m.Error = execute(m)
+	m.End = time.Now()
+	select {
+	case send <- m:
+	case <-ctx.Done():
+		return
+	}
 }
 
-func (b *benchmarkThread) execute(measure benchMeasure) {
-	if !b.interactive {
+func execute(m benchMeasure) error {
+	if !m.interactive {
 		// do not use parent context that will cancel in flight requests
 		// because we give test some time to gracefully wrap up
 		// the in-flight connections to avoid extra errors
-		measure.Error = b.client.SSH(context.TODO(), nil, false)
-		measure.End = time.Now()
-		b.sendMeasure(measure)
-		return
+		err := m.client.SSH(context.TODO(), nil, false)
+		if err != nil {
+			return err
+		}
+		return nil
```

Comment:
```suggestion
		return m.client.SSH(context.TODO(), nil, false)
```

Reply:
I think we should keep it at is (I am going to revert it back because I made the change) because we still need to execute the command and exit out of the session with `writer.Write([]byte(strings.Join(m.command, " ") + "\r\nexit\r\n"))`. And it doesn't make sense to write the command before we ssh.

Reply:
Hmm, `writer.Write` is only for interactive commands.
It looks like this code is not calling any command here at all. This line should be something like:
```
return m.client.SSH(context.TODO(), m.command, false)
```

Reply:
I am dealing with non-interactive sessions on lines 244-253. So if we get passed the conditional `if !m.interative { ... }` then we will be dealing with interactive sessions.

Reply:
Hmm, maybe github is showing us different code spans.
I highlighted this comment on the block (244-253):
```
	if !m.interactive {
		// do not use parent context that will cancel in flight requests
		// because we give test some time to gracefully wrap up
		// the in-flight connections to avoid extra errors
		err := m.client.SSH(context.TODO(), nil, false)
		if err != nil {
			return err
		}
		return nil
	}
```
which is for non-interactive sessions.

The original comment was that it's equivalent to:
```
	if !m.interactive {
		// do not use parent context that will cancel in flight requests
		// because we give test some time to gracefully wrap up
		// the in-flight connections to avoid extra errors
		return m.client.SSH(context.TODO(), nil, false)
	}
```

But I also notice that the 2nd parameter to `m.client.SSH` for non-interactive sessions is `nil`, but should be `m.command`.
Or is the command for non-interactive sessions passed somewhere else to `m.client`?

Reply:
Oh yes, I only saw a portion of the code span, that is my bad. You're right, returning like you suggested above will work